### ISSUE TITLE
Query parameters lost on redirect to OIDC provider

### ISSUE THREAD

**Issue by manuelbl** [quarkusio/quarkus#9539]
**Describe the bug**
When using OpenID Connect (OIDC) to protect web applications, query parameters are lost if the user is redirected to login first.

**Expected behavior**
After the user has successfully authenticated with the OIDC provider, he/she should be redirected to the same URL that was initially requested.

**Actual behavior**
After the user has successfully authenticated with the OIDC provider, he/she is redirected to a different URL. The URL has the same scheme, host and path but different query parameters. The original query parameters are missing. Instead the query paramters `session`, `state` and `code` are present.

**To Reproduce**
1. `git clone https://github.com/quarkusio/quarkus-quickstarts.git`
2. Open the project `security-openid-connect-web-authentication-quickstart `
3. Configure a valid OIDC provider in `application.properties`
4. Extend the class`TokenResource` as shown below
5. Open the URL `http://localhost:8080/tokens?myparam=abc`

After authenticating, the URL is something like `http://localhost:8080/tokens?state=8b...3e&session_state=3f...57&code=bc...5c (instead of `http://localhost:8080/tokens?myparam=abc`).

The page starts with `myparam: null` (instead of `myparam: abc`).

**Configuration**
Same as example project except for:
- `quarkus.oidc.auth-server-url`
- `quarkus.oidc.client-id=abcfund`
- `quarkus.oidc.credentials.client_secret.value`

**Modified code for `TokenResource`**

Add query parameters in signature of method `getTokens` and output them as HTML:

```java
    @GET
    public String getTokens(@QueryParam("myparam") String myparam) {
        StringBuilder response = new StringBuilder().append("<html>")
                .append("<body>")
                .append("<ul>");

        response.append("<li>myparam: ").append(myparam).append("</li>");
        ...
```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Microsoft Windows [Version 10.0.18363.836]
 - Output of `java -version`: openjdk version "13.0.1" 2019-10-15
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.4.2.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Apache Maven 3.6.2 (40f52333136460af0dc0d7232c0dc0bcf0d9e117; 2019-08-27T17:06:16+02:00)

**Additional context**
The OIDC provider used was a Keycloak instance configured for authentication code flow and requiring a client secret for the flow.

**Additional thoughts**
I'm asking myself what would happend with a POST request. Would the request body be saved and replayed after the redirect?

**Comment by sberyozkin** [comment]
@manuelbl Thanks, yes, it is a known issue, I'll get it fixed for 1.6.0.

> I'm asking myself what would happend with a POST request. Would the request body be saved and replayed after the redirect?

If we have a code flow then it is always a human user who is trying to access a web application via GET first. Once the user is authenticated, the follow up POST requests will work fine. 

Otherwise `POST` would only work with the bearer tokens.

**Comment by tarundvarma49** [comment]
Losing query parameters during a redirect to an OpenID Connect (OIDC) provider can occur due to various reasons. Here are a few potential causes and suggestions to address the issue:

1. Incorrect redirect URI: Ensure that the redirect URI provided in your OIDC request matches the one registered with the OIDC provider. If the redirect URI does not match exactly, the OIDC provider may strip the query parameters for security reasons.

2. URL encoding issues: If your query parameters contain special characters, make sure they are properly URL-encoded before appending them to the redirect URL. Incorrectly encoded query parameters can be misinterpreted or stripped during the redirect process.

3. OIDC provider configuration: Some OIDC providers may have specific settings that affect the preservation of query parameters. Check the configuration options and documentation of your OIDC provider to see if there are any relevant settings or restrictions.

4. Backend server configuration: If you are using a backend server or reverse proxy to handle the OIDC redirection, check its configuration. Some server setups or proxy configurations might modify the URL during redirection and unintentionally remove the query parameters.

5. Validate OIDC implementation: Double-check your OIDC implementation to ensure that you are correctly following the OIDC specifications and best practices. Make sure you are using the appropriate libraries or SDKs for your programming language or framework.

By reviewing these possible causes, you should be able to identify the reason behind the query parameter loss during the redirect and take appropriate actions to resolve the issue.
